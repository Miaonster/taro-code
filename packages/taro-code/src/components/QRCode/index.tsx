import React, { CSSProperties, useMemo } from 'react'
import { Image } from '@tarojs/components'
import { createQrCodeImg } from '../../common/qrcode'

type TypeNumber =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | 32
  | 33
  | 34
  | 35
  | 36
  | 37
  | 38
  | 39
  | 40

const QRCode: React.FC<{
  className?: string
  text: string
  size?: number
  scale?: number
  style?: CSSProperties
  errorCorrectLevel?: 'L' | 'M' | 'Q' | 'H'
  typeNumber?: TypeNumber
  foregroundColor?: string
  backgroundColor?: string
}> = ({
  text = '',
  size = 100,
  scale = 4,
  typeNumber = 2,
  errorCorrectLevel = 'M',
  style = {},
  foregroundColor = '#000000',
  backgroundColor = '#FFFFFF'
}) => {
  const image = useMemo(() => {
    const options = {
      errorCorrectLevel,
      typeNumber,
      size: size * scale,
      black: foregroundColor,
      white: backgroundColor
    }
    return createQrCodeImg(text, options)
  }, [
    errorCorrectLevel,
    typeNumber,
    size,
    scale,
    foregroundColor,
    backgroundColor,
    text
  ])
  const widthString = size != null ? `${size}px` : ''
  const heightString = size != null ? `${size}px` : ''
  const finalStyle = { width: widthString, height: heightString, ...style }
  return <Image style={finalStyle} src={image} />
}

export default QRCode
