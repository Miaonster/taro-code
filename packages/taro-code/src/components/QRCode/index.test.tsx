import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import QRCode from './index'

describe('QRCode test', () => {
  it('hello world', () => {
    const wrapper = shallow(<QRCode text='hello world' />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('40+ words', () => {
    const wrapper = shallow(
      <QRCode text='12345678901234567890123456789012345678901234567890' />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
