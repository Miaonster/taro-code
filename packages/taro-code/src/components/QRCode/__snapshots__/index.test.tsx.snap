// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCode test 40+ words 1`] = `ShallowWrapper {}`;

exports[`QRCode test hello world 1`] = `
<ForwardRef
  src="data:image/gif;base64,R0lGODdhkAGQAYAAAAAAAP///ywAAAAAkAGQAQAC/4SPqcvtD6NcodqLwwS5+19t4EiW4TaZJ8q27gvH8kxTqqjmeM6XdaL7CYfEovGoCEp6ph3zmSkqkdSq9UqdQqAjJ5crvWHH5LI5on18P941M9w8y+f0atrh7rTz91m/DhgouCRGyGexd+hD9Dfo+DjYmKS4gkbJA7cIuckZKAl0mXjJxljYeYo69okwKjqqVxqXOktrtHrQmvIqK3Rb+wu8ZSpM6bp7kUkSvMyM4ssRqnvcFavZfI0NykusaDytUa2cPZ79nGv47ZFMTd6+bB6Nnh4Vzu4+N58PLq9fyX24DoQ3ZNL6HQtosFjBhAPf1BO40N6/hPF6UdzVEGPEPP8I1W0kxe8iwIciOX7MlxETSY8hQVoqqXAIzIoT+6V8ljLnyZlQOvJ8cpOmGqE04G0b+tOkzKRKX1IMerToMKTWajJ1uPTqF6jddoJZCaslS6dagYItW5UqQ689z2LQKRZtWqlysZI1yFVc1rkNjNZVuffvVDwi80qs4VfvXcGKfzAGvFifYYiBDxOO2vfxYD+aMTMoOdml482fSdvo3Bgx6tSXn7I1W5ly5LFWVxOMbfv27Hmhaav2fJpv6dy+6RKnFzdd77AWgU9yru34W7e6zSS2rKowPtPPhV/B+RrRduisuNvRLue6bOvmXYDHXTwzeSSgx3vvzjr7fBbvm9//j54fFvWl1x4uBR7RX23MsbcfNA36pJx9AQY3oRUJqoWdgAc6+F8W6J2hnmhlXNhahec1SCJnF9ERYnwaPhiePyNumKIMA4JI44YQTgffgjOiqCN1B0mYIYVFWhhkcjziCCSMPUZIYJMd0pfkbsgxOaUBNcZwI4NZcmiih05auaSXYRo45mgrErkehm2SsSV+R574ZZwvdPljnVWq+SVcWJ6p5Z6/uRalnmkOCqgzghqXqJ3ufWjmnOUdyqikLTh6aY6Uqkjon5YGuqmNi8rXp5AylijXjtW5KVmM+3j6JqeJqiqekj/Reiqpa9laK6wiVhrri4b+heurCkJJ5qpw/446XKlPlsmqVsUu9xWvuQo7q6ssakqsqcZGy5u2bP4q66e2cFvXtK6ek2yvkQbLJbPxSpmut9S2Ze23y4ZqJLxI0puqvesSpau56j6rrH7D1ouwu+Aieyy0+zrb7rXfoRtwwxbLaZO4hWab77YAo3Vwxa/4+S65ovILQ4tXlRzxNCjnCbLJIi+csX+NQvqwxArXHLPDE+8sGMw9y+yxry7OSzGfRHersb4FdxzyuEu3LC/WI5dlNKpU22y1j8CqfPHWa9ybMCTsBr0xnUAfLfTP5q4Ndw+00O012f82XbfU2M5NcLMjpYL31Hq7bXDSKV8NIB9ox6124P0eLibfef8zXvnTg5uMKYKSc0w5lSw37i/ipU/aVb6dnxsT2B8nXvXrp6OZOudZm+7GzEPD7rrSYht+9sCjZ+644jTzzrbfZePcVPKrF+vylYv/vrLl5c4Oau3O3058833f/Pbl1C+vefGqcy+69t+HLT2iyLuPPZjm2z58+psnD/774re/e/yFCz6/U/xvcpizn/6AN769lc97+8MXKgYIugJ6rn7ZC9/1QlfBAPbted6Kns/+Fj/dyc1/nyNd7mYBQRNKkHXWA6B0lBfBBEaCZw2EDf2gprMXwlCFMhQEnmpol+0VrYOMEWHk1nQ+sxEHekU0Hid+iMBq3ZBhOXyhER8BRRf/njCJOHSaDq/oiCwScCvC6yL8pAPGGSJxioDTIeSY5sY0+pCGUXSgEM04tuPI0RN01GLwuEhFL1rRiZsQYwyleMdAnlGPhDxipxLZRjfu8FGSnCTtLAjCVgEyknEk4hBj98A+jhGRG2xi1B6zR0AYkoekBGJnmPjJ3glQlIe0YyljKUg0NhKLtGSlLV2pmXvUwYNpG+UHjVnMWibTl8sUpjP5qET+1bGZl+SkNUP4zGz2D4PyI2E0j6nMN4azbdos5/G8yTxuElOczGSnOd9JvmuqE2PoXCA34YlPgYESf7scpyUziM18CvSCK1wnOatZT3mucKAMhaNCe4jQeX6T/5oAvWdDL8ofeko0nQXV6EIxCtJ9rk+kwExlNy0a0pRO050R7ehEWVpRl7ZQpSk16D9PKlN75vSANK2pRyEaU6DiVKg27alRM/XSgwZVmn4k6k+ZetSo+tOkRe1nO5UqVaNWlaQrxepWs7rUSr7Mk+rbpljPitY/VlGTvkurW986U3+msHtwratdhfo4u7Hvrnx9KyzLOsK+CtavZL3fOQeL2LP+1bBmTaxjB3nKEtL1sZRdYmE12NjKajaYl2VgYDcLWlOuFSV7Da1pxxrZ1k3vtKxNymIx+9nWynYmr/VsJmeL215S8pGHza1v8dLZLbb1t+C86k1XuVvGdhUyuf+UVhl1ilerom5XoyXjJgPKxq9BMqHY5ed1hdtc62YXpeMl7Xd32t2Rlle84W3lcleHyuc+FKpa42i4Umtf91ZPkcD8anuLO13g4he6+nUoyeTLXfJ6d70Frq9ym9rgO4l2uxvFJDJhOlSILZK9t4yrcZHbUg6XVJ8M/iVBRfxe9IX1VgiuME8v7FXdIhWwEDbxfnNGYfQqWL05RvGNbQtj5m6YKXk9roozDOAQaziPPg4yfCfcYQInOaNJPVlwdazW/7L4vE6lIJIp+uUhDXi+NjYw11qM5ehytcZIG3OCszxk16K5yx4O8JvvW90I2xm8cd5yicH8ZNVGGbZsxvD/XA2o3T5j2KRFLrJ/B51eJi/5xPSVsJcPvWIxyxLSO5b0N7pG6SkrStD9laySRY1pFlJXy4uW7owf7GRTZzrGpI7tpH/c6eXW1sWJHjGXKz1ZPLMaq4yeM7BP3eZNl9rLjj5yoOuM7CByGs6+/nOr3cxnT4O52L9GtYxnbeRa31bYiia2q6lM41iLO8iNLq2mh31Tblub1rytdo+PPcF6h3rb5x51uuUq6zDTG9YKFHCeRS3ve3t7jQtWOL+xTe19X1vZEtdrt23Y8Gmr+eDJhjeocR3xQmtkzeqGNrpXXe5497vdF5e2vTWOb1Ub3OMkpnnLzU3ygt9Zz9EeOcVB/77xjC/b5AJXedRYPm8hv1zbExd6ii99ZKoGHMTg/riZ08zzk7/YwWTGOMzLPNWapxzpNh8uoKf+7aJbnes7B7uliS71dQNc7sZd+9ul7HWm45zjhkY7w3k8drEP3cJhlx2vs/30f9dd8BUPt+KTu/W7d93liSf44o/OeJErvfH5a3veK09odmf+6nS2POQjzXljL1zfmrf76Q8fctLH3PU937zmU/1swl8+6bTPN94pf3u/Ez33k7c43/dedof3/e9LTz3mb270n8fT88Avuelr33t//974ye9t6UM/98djf/Q6zzXQY57qzrM27sOPer/T7+7Ksl/3e/7o83mPQv/fzj/y9Q964KF/N/r3fs7mfjn3YQV4fJS1f6jXevHneNenSgJogP03eytHfgfIbFdmVwtofqJneFjndnOUWxxof9JXfiD4eY40WyTof803bugXcOp3WixYgROIaLDXZGEkgSa4e92XWasHgcO0g07XgPe3WkAIfmqEWzSIcBZohF/3cGanWUwYhUQ4fTiYdUI4ggMIdRlog+DGQRC3gVyYXyE4fmIIg3T3eiVoheG3fbYne2ZYdRqYhuKndQzogfgHgHSIhHjohsUHh2zXgYWnh4UIb4+md4GIge3XhWWYgtbHf7WHiI1nU8Tnhz0Ihctnh9qXhBTYhHsIigmYfbj/F4Oqd3aNGITw94WcSH3cl4nI94oPGIlg+FRyyIqDiImDB4iWiHhFmICVSICouIu1+IiCGHuEeIjEWH3G2IK6ZoqaGIRkB3iJCIzCeImeeIqbqIqGOI2UqIyumIijKHzR+IwDN4tFV42MqI5vmI70J4nfGH3KZ45Axow1yIPYuIyL6I60WGW7hoL5uIbN6IvyKIv0KHkGiY+wqIutqIgBaY9tqI/DWGWP5Y85GJEIeYaieIEJqYB8WIwX2Yt56IPOl3SK5ZEA+Y5qmJEjeX59uH4nCY6gh5Fz+IThWI5oVZFZSHVqt5GvJpGIlZO2uJO8yIbdGHyJFZQfyZGkaI24/8iRzVZJSYmS/KiNwbiOPzlYUhmTR1mVTVmULgiJfaWVDZmSXXmV17iSsWiSaCiUaUeUAimSailWY9mOnZiWNlmTMslXdAmPTHmWTnmXenlXfDmR43iOb1mHKwiTZEmVqWiV+8iT0LeWv2iYveaMedmSVeh9txaH2UiOmBmXCwmYNBlKzEeSBCmD38eQiPmJUuhzLBmWo3mL/ziPIVmPrXmEHfd/3PiDnomVh4mAyaiSd4iFtiiOhemVcHmC5IaXvGlrLkmbjDmbyok75qWRoZib0PibaEmas2SamamdMylzDCmNzemdrNeZ4Wmb1amaxamUPsmd48lWsPmH8UmF6v95jAcpm4T5l9EZhvcYmFwJmf8pnJt4n1t5hdFZnqfZCTt5oNJ5nI44ldP5kJv5aT3ZmOf4oAsKngr5nLppnrGYmom5nbJJoCm3jRC5lMFplKEpmNT5lDc5onCVfWM5o6txo25Vo4uZo6/kgJNJn+yZhfyJcr2poxgqn/kZbLnRoziJpDb6o1BmpGm1o2ypmdQopR96pKC5nEO6mA76pVnaofGIoiwqms71gWPIpUJqnGH6nQmqptfpmhDqpujZpTT6pDwapbhkoYS1pkvaplbKE00KpLuZnQh6pzi6p5JUpZTphSWZdkSKY4bapxMqqZb5gnj6pzeopJzKnN74psj/KJfDSaG4OQ6TOKZD2ZeR6oTtgKrpSVtkyF8Mmk+vept+pqJlWaS0ik+2qp9oCqAZiqmwSqblAI+Xup4xKpmluIrC5KsOKWerGF+b+kzPCp9nJq1iSqwYZa3ECay5KqzzSam1eqx16pjIaae/CqPA0K2liqstWp/iGqLk2o/mapfdOaoG2qrk0K5QaZbsyKrNeg/9epN+CbChqq4Vaqz1KqivCa8gaZ1Bqk0Eu6ykWpfDeqtNd6qrSlz9+akqOHOVWpsdq7Av6qVpypmJip8ke6UmG6hzmqLj6rKbhaxfqaUOO6UFybKmeqYWKYLpCqi+ubM8O7OWuqifKacESZH2/2qfTLufsjq0RCug96qFQOupJBq1GtuzJwuzzJq0+ZqVTmuzmQqiN6uzWVuz6+p7IZuz/hq2DUunKHuhciu0WVusRYuoSrirZLuyQ5u2Jcu3OGu2bitYfyu1cJqybFq3aCu2apukH4u4h9tTF+ue7yazAwm2GnpRlOufXguukQmpjzpQnNueEXu5Ljq14glWyuqcqZus16qgBWumvcqxEiq4Y/qgMStVpIu18sqrmLu1LwtWvAudkJuxI1u6xru6wIu3Z6u80Lq4yZu4yyuqwRu90wu70lu82Eu9ARqbnUuqGyq7oitQxHu9c/u1YBmvndq9+Gq9fYu+Euu96+uz7f/rvs1LuPHZrmC6vOYLv2WLu1CLsfY7vxCrvQC8rQLsuwSci/g7vh7bu+CrrxVrteHKvuhYuwf7r1RruAhMvzpZma9rwfU7woBLrNMaurYrnQbLtg5Mt6iRv5WbtyuKruc6wY6KR83rv8gLwghLw3D3wHv7vSSMwTVcohesu6Caw66LxCHcxD6sq5pLwUL8wW1psU78vFHctFPcwkxMxCwsnqoavo3LvUW8wWGMxWWcxKmqrVX8niW8vVZcwWYMmTFsuu87w1oswyvslp5bpnyKx3x8xYM8x2t8wm1swD18xiJMx1RryMfLWVy8mhl8xF/sx1gKyDpMyZNsxGg8xtT/+rgRTMQBSL4Lq8IJd7purLWFNLvXsL+tjLpDbMI6WMrY8Mq1zLxePMt6K8WuWq4CC8qPzMtb7MsMG6yhHMdvXAsnaspviMrz6sLFDMTucMvTfMzC/LN1PLC/fMwRirTnqc3UzM2f680cnH+43AzVHM7XfMmsjM7MoM69/LnYDE3WzK/j/LBB+79GSziwDM/4rL6Ku895/AvMbMsAfZlzyqHp7M/vgNC5HLnnq8ykbM8ba8zkjKRirMI7bGWSzMgGDdGqTNCgi5pB7MiIzLqH6rxqvMl7rNGzGtICzcPxm896LMFUnMg0HdD6PNMe3MC6nMyDatKeXNE7bdOW3Mku/93HmRzTPL3SOp3QJe3RT7zEsayyPW25Rg3HEi3UU43URR3VmUueMjrUH43SBezUdnzAAy3IVS3SHd3NGb3UG93Sd6zJhLzO8zzXGszXXSzLUN3UVxvUt2vVOc3VserVipzXNb3VE8rRaq3Yk6rVyHzYcK3XeC3Zd73IY5vAiT3YT63MKKy0aSy5CfvZkF3aE9vQ2SzPjE3SYv204vzOwxzb7JzUF8ytq13Piz3ZHA3SDD3brE3Ml03Xn8xQv03LYB3Y9wvUE725uh2BwY3W093YI53b0h3dyl3YKQ3bnG1OyE3b3m3alY3aLQtP4C3ctU3cfa26z43dVcvbYd3bAf973NAN360937c9yu6t3ckd38v92oGM3uxq301q4Aps10c14Nn939tN3Y3c3g214Pc93K7t2wV+zv0d3paFna6duxgOtxcOxdUN2qsM4PTM4H791i990skpylwLzipu2F3dug7urWtt3RmO0z8M039N3nJt3DE+wDye2c3N1uUc4aVZyHNJ1sBs2zcs5Avs4zQu1XGdrUHeoHv9RU1u5U/+zVk+4hCOrTVevb875R4KskN+1ETG5Rh95VAO5kselW3u4Qi+3/0c5oUayM8s39F8z3nO5B0+2R/+3h3M1qItovtq4e2c5lK+4lpO1ANa145LqIz+ThOO4mJ+05UM41mdhemW/tYL/eiFvs3vnelgbNZh3r2fjuV7TucCDuLS3OChrqeCrtmzXk6sDueubutGnuP9a9+nTtqhTd/DG+ygPuNYPdL8u7q6/uW8TubJHrfNfuyt7ueJ7tnmrVXVvuvXDu2jTeqyreFErr/D7tjF7unc/uzezu6+Pu3GburITu5Krd+dvrvqbs69fuYlTtgMzNwRzc/RPu+bLp4FAAA7"
  style={
    Object {
      "height": "100px",
      "width": "100px",
    }
  }
/>
`;
